language/move-vm/types/src/values/values_impl.rs

```diff
@@ -476,30 +505,50 @@ impl IndexedRef {
+            | (General(v1), Resource(v2))
+            | (Resource(v1), General(v2))
+            | (General(v1), General(v2)) => v1[self.idx].equals(&v2[other.idx])?,
+
             (U8(v1), U8(v2)) => v1[self.idx] == v2[other.idx],
             (U64(v1), U64(v2)) => v1[self.idx] == v2[other.idx],
             (U128(v1), U128(v2)) => v1[self.idx] == v2[other.idx],
             (Bool(v1), Bool(v2)) => v1[self.idx] == v2[other.idx],
             (Address(v1), Address(v2)) => v1[self.idx] == v2[other.idx],
 
             // Equality between a generic and a specialized container.
-            (General(v1), U8(v2)) => *v1[self.idx].as_value_ref::<u8>()? == v2[other.idx],
-            (U8(v1), General(v2)) => v1[self.idx] == *v2[other.idx].as_value_ref::<u8>()?,
+            (Resource(v1), U8(v2)) | (General(v1), U8(v2)) => {
+                *v1[self.idx].as_value_ref::<u8>()? == v2[other.idx]
+            }
+            (U8(v1), Resource(v2)) | (U8(v1), General(v2)) => {
+                v1[self.idx] == *v2[other.idx].as_value_ref::<u8>()?
+            }
 
-            (General(v1), U64(v2)) => *v1[self.idx].as_value_ref::<u64>()? == v2[other.idx],
-            (U64(v1), General(v2)) => v1[self.idx] == *v2[other.idx].as_value_ref::<u64>()?,
+            (Resource(v1), U64(v2)) | (General(v1), U64(v2)) => {
+                *v1[self.idx].as_value_ref::<u64>()? == v2[other.idx]
+            }
+            (U64(v1), General(v2)) | (U64(v1), Resource(v2)) => {
+                v1[self.idx] == *v2[other.idx].as_value_ref::<u64>()?
+            }
 
-            (General(v1), U128(v2)) => *v1[self.idx].as_value_ref::<u128>()? == v2[other.idx],
-            (U128(v1), General(v2)) => v1[self.idx] == *v2[other.idx].as_value_ref::<u128>()?,
+            (General(v1), U128(v2)) | (Resource(v1), U128(v2)) => {
+                *v1[self.idx].as_value_ref::<u128>()? == v2[other.idx]
+            }
+            (U128(v1), General(v2)) | (U128(v1), Resource(v2)) => {
+                v1[self.idx] == *v2[other.idx].as_value_ref::<u128>()?
+            }
 
-            (General(v1), Bool(v2)) => *v1[self.idx].as_value_ref::<bool>()? == v2[other.idx],
-            (Bool(v1), General(v2)) => v1[self.idx] == *v2[other.idx].as_value_ref::<bool>()?,
+            (General(v1), Bool(v2)) | (Resource(v1), Bool(v2)) => {
+                *v1[self.idx].as_value_ref::<bool>()? == v2[other.idx]
+            }
+            (Bool(v1), General(v2)) | (Bool(v1), Resource(v2)) => {
+                v1[self.idx] == *v2[other.idx].as_value_ref::<bool>()?
+            }
 
-            (General(v1), Address(v2)) => {
+            (General(v1), Address(v2)) | (Resource(v1), Address(v2)) => {
                 *v1[self.idx].as_value_ref::<AccountAddress>()? == v2[other.idx]
             }
-            (Address(v1), General(v2)) => {
+            (Address(v1), General(v2)) | (Address(v1), Resource(v2)) => {
                 v1[self.idx] == *v2[other.idx].as_value_ref::<AccountAddress>()?
```

Comment:
We're getting a bit bloated here, but I'm not too worried as `or_patterns` has landed. Hopefully we'll be able to bear with this until the day `or_patterns` gets stabilized ;)